File: bindings/javascript/src/caldav.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// CalDAV protocol types for proven-servers.

/** ComponentType matching the Idris2 ABI tags. */
export const ComponentType = Object.freeze({
  VEVENT: 0,
  VTODO: 1,
  VJOURNAL: 2,
  VFREEBUSY: 3,
});

/** CalMethod matching the Idris2 ABI tags. */
export const CalMethod = Object.freeze({
  GET: 0,
  PUT: 1,
  DELETE: 2,
  PROPFIND: 3,
  PROPPATCH: 4,
  REPORT: 5,
  MKCALENDAR: 6,
});

/** ScheduleStatus matching the Idris2 ABI tags. */
export const ScheduleStatus = Object.freeze({
  NEEDS_ACTION: 0,
  ACCEPTED: 1,
  DECLINED: 2,
  TENTATIVE: 3,
  DELEGATED: 4,
});

/** CalError matching the Idris2 ABI tags. */
export const CalError = Object.freeze({
  VALID_CALENDAR_DATA: 0,
  NO_RESOURCE_TYPE_CHANGE: 1,
  SUPPORTED_COMPONENT_MISMATCH: 2,
  MAX_RESOURCE_SIZE: 3,
  UID_CONFLICT: 4,
  PRECONDITION_FAILED: 5,
});

/** ServerState matching the Idris2 ABI tags. */
export const ServerState = Object.freeze({
  IDLE: 0,
  BOUND: 1,
  SERVING: 2,
  SCHEDULING: 3,
  SHUTDOWN: 4,
});

File: bindings/dart/lib/src/caldav.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// CalDAV protocol types for proven-servers.

/// ComponentType matching the Idris2 ABI tags.
enum ComponentType {
  vevent(0),
  vtodo(1),
  vjournal(2),
  vfreebusy(3);

  const ComponentType(this.tag);
  final int tag;

  static ComponentType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// CalMethod matching the Idris2 ABI tags.
enum CalMethod {
  get_(0),
  put(1),
  delete(2),
  propfind(3),
  proppatch(4),
  report(5),
  mkcalendar(6);

  const CalMethod(this.tag);
  final int tag;

  static CalMethod? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ScheduleStatus matching the Idris2 ABI tags.
enum ScheduleStatus {
  needsAction(0),
  accepted(1),
  declined(2),
  tentative(3),
  delegated(4);

  const ScheduleStatus(this.tag);
  final int tag;

  static ScheduleStatus? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// CalError matching the Idris2 ABI tags.
enum CalError {
  validCalendarData(0),
  noResourceTypeChange(1),
  supportedComponentMismatch(2),
  maxResourceSize(3),
  uidConflict(4),
  preconditionFailed(5);

  const CalError(this.tag);
  final int tag;

  static CalError? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ServerState matching the Idris2 ABI tags.
enum ServerState {
  idle(0),
  bound(1),
  serving(2),
  scheduling(3),
  shutdown(4);

  const ServerState(this.tag);
  final int tag;

  static ServerState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
